From ab0f7a6e1bac5fb0ebb18cbdfaa0ff1d5d7eed14 Mon Sep 17 00:00:00 2001
From: Ghennadi Procopciuc <ghennadi.procopciuc@oss.nxp.com>
Date: Wed, 1 Sep 2021 13:47:59 +0300
Subject: [PATCH 15/17] gpio: siul-s32gen1: Share eirq regmap among siul2
 instances

commit ab0f7a6e1bac5fb0ebb18cbdfaa0ff1d5d7eed14 from
https://source.codeaurora.org/external/autobsps32/linux

This commit makes eirq regmap shared among siul2 instances
to avoid the cases when one instance overwrites the values set
by another.

Upstream-Status: Pending

Issue: ALB-7606
Signed-off-by: Ghennadi Procopciuc <ghennadi.procopciuc@oss.nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/gpio/gpio-siul2-s32gen1.c | 30 ++++++++++++++++++++++++++++--
 1 file changed, 28 insertions(+), 2 deletions(-)

diff --git a/drivers/gpio/gpio-siul2-s32gen1.c b/drivers/gpio/gpio-siul2-s32gen1.c
index 8a10b3a5d085..04252909ad14 100644
--- a/drivers/gpio/gpio-siul2-s32gen1.c
+++ b/drivers/gpio/gpio-siul2-s32gen1.c
@@ -66,6 +66,8 @@
 
 #define SIUL2_0_MAX_16_PAD_BANK_NUM	6
 
+#define EIRQS_DTS_TAG	"eirqs"
+
 /**
  * enum gpio_dir - GPIO pin mode
  */
@@ -494,7 +496,8 @@ static const struct regmap_config siul2_regmap_conf = {
 };
 
 static struct regmap *common_regmap_init(struct platform_device *pdev,
-					 struct regmap_config *conf, const char *name)
+					 struct regmap_config *conf,
+					 const char *name)
 {
 	struct resource *res;
 	void __iomem *base;
@@ -632,12 +635,35 @@ static bool irqmap_volatile_reg(struct device *dev, unsigned int reg)
 static struct regmap *init_irqregmap(struct platform_device *pdev)
 {
 	struct regmap_config regmap_conf = siul2_regmap_conf;
+	static struct resource *glob_res;
+	static struct regmap *glob_reg;
+	struct resource *res;
+	struct regmap *reg = NULL;
+
+	res = platform_get_resource_byname(pdev, IORESOURCE_MEM, EIRQS_DTS_TAG);
 
 	regmap_conf.writeable_reg = irqregmap_writeable;
 	regmap_conf.volatile_reg = irqmap_volatile_reg;
 	regmap_conf.val_format_endian = REGMAP_ENDIAN_LITTLE;
 
-	return common_regmap_init(pdev, &regmap_conf, "eirqs");
+	/**
+	 * For the cases when the same EIRQ block is shared among
+	 * multiple instances
+	 */
+	if (!glob_res) {
+		reg = common_regmap_init(pdev, &regmap_conf, EIRQS_DTS_TAG);
+		glob_res = res;
+		glob_reg = reg;
+		return reg;
+	}
+
+	if (glob_res->start == res->start)
+		return glob_reg;
+
+	if (reg)
+		return reg;
+
+	return common_regmap_init(pdev, &regmap_conf, EIRQS_DTS_TAG);
 }
 
 static bool not_writable(__always_unused struct device *dev,
-- 
2.17.1

